AdminEmailLogPage and AdminActivityLogPage fails when there is a patch version #4417

bamboo3250 opened the pull request:
Fix #4417
@damithc @wkurniawan07 
My solution is using `getVersions()` method of `ModuleService` to get the list of available versions and select the current version together with its preceding versions (at most three). The only problem now is that I'm not sure about automated testing this function on both local and staging server.

wkurniawan07 commented:
Can you revert the header comments to unrelated methods? Sure it's great to have them, but this PR is never about those.

bamboo3250 commented:
@wkurniawan07 okay, no problem. Done.

thyageshm commented:
@wkurniawan07 @bamboo3250 do you guys think this one can be merged this release? If yes, please add the milestone to the PR and the issue :-)

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -225,64 +228,49 @@ private LogQuery buildQuery(AdminActivityLogPageData data) {
+    /**
+     * Selects versions for query.
+     * @param versions
+     * @return
```
Can be simplified these three lines into one sentence `Returns ... for versions ...`

damithc replied:
I don't think the current comment is a complete description of the method though.

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -225,64 +228,49 @@ private LogQuery buildQuery(AdminActivityLogPageData data) {
+                    } else {
+                        return defaultVersions;
+                    }
+                }
```
This `for` look looks like it has too many nesting levels for such a short code block. See if you can simplify.

damithc commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -23,7 +25,8 @@
     private static final int MAX_LOGSEARCH_LIMIT = 15000;
-
+    private static final int MAX_DEFAULT_VERSION_NUMBER = 6;        // 6 default versions for query.
+                                                                    // include the current version and its 5 preceding versions.
```
Write a proper javadoc comment instead?

damithc commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -81,59 +82,47 @@ private LogQuery buildQuery(String offset, boolean includeAppLogs, List<String>
+                    }
+                }
             }
         }
```
This `for` loop looks familiar. can reuse?

bamboo3250 replied:
Yes, this is what I proposed in #4773. I'll refactor it there, maybe not in this PR.

bamboo3250 commented:
@damithc I fixed the comments and simplified the logic in the for-loop. Please check if they look good to you or not.

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -33,11 +35,15 @@
+    /**
+     * 6 default versions for query, including the current version and its 5 preceding versions.
+     */
+    private static final int MAX_DEFAULT_VERSION_NUMBER = 6;
```
A more meaningful name, such as `MAX_PAST_VERSIONS_TO_QUERY`?

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -225,64 +231,45 @@ private LogQuery buildQuery(AdminActivityLogPageData data) {
+                defaultVersions.add(version);
+                if (defaultVersions.size() >= MAX_DEFAULT_VERSION_NUMBER) {
+                    return defaultVersions;
+                }
```
Seems like the code is doing two things. It is more efficient that way, but less readable.
Can you break this down to two steps? e.g.
```java
int currentVersionIndex = getCurrentVersionIndex(versionList, currentVersion);
return getNextFewVersions(versionList, currentVersionIndex, MAX_DEFAULT_VERSION_NUMBER);
```
BTW, you are using a `Set` as it it is a sorted list. Is that correct?

bamboo3250 replied:
A `Set` is not necessarily a `SortedSet`. It is what `getVersions()` of `ModulesService` returns. I can change it to `String[]` so that we can have more control of the list.

damithc commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -23,7 +25,12 @@
+     * 6 default versions for query, including the current version and its 5 preceding versions.
+     */
+    private static final int MAX_DEFAULT_VERSION_NUMBER = 6;
+    
```
Can we avoid this duplication?

bamboo3250 replied:
It could be better if we move it to Const class.

damithc replied:
Or create a `Version` class?. `Const` class cannot absorb duplicated methods.

bamboo3250 replied:
Then I would like to resolve #4773 in this PR as well since they are closely related. Is that fine?

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * Finds the index of the current version in the given list.
+     * Returns the size of the list if the current version is not found. This is not supposed to happen!
```
If it is not supposed to happen either put an `Assumption` or throw a runtime exception?

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -155,15 +147,16 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
             
-            LogQuery query = buildQuery(data);
-            List<ActivityLogEntry> searchResult = searchLogsByQuery(query, data);
+            //LogQuery query = buildQuery(data);
```
commented line

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -173,116 +166,36 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
+        
+        for (AppLogLine appLog : appLogLines) {
+            String logMsg = appLog.getLogMessage();
+            if (logMsg.contains("TEAMMATESLOG") && !logMsg.contains("adminActivityLogPage")) {
```
Introduce variable to explain meaning of condition?

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -173,116 +166,36 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
+                    appLogs.add(activityLogEntry);
                 }
-            }    
+            }
```
you can get rid of most of this nesting by reversing the condition and using an early termination. e.g.
```java
if(!condition){
    continue;
}
```

damithc commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -59,123 +52,29 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+                    currentLogsInPage ++;
                 }
-            }    
+            }
```
Another place where reversing the condition can reduce nesting?

damithc commented:
Added some comments. Once done, I'll wait for `toMerge` for final review.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * 6 past versions to query, including the current version and its 5 preceding versions.
+     */
+    public static final int MAX_PAST_VERSIONS_TO_QUERY = 6;
```
If it includes the current version, maybe a better name is `MAX_VERSIONS_TO_QUERY`?

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+    public void setQuery(List<String> versionsToQuery, Long startTime, Long endTime, String offset) {
+        query.includeAppLogs(true);
+        query.batchSize(1000);
+        query.minLogLevel(LogLevel.INFO);
```
Consider making `1000` and `LogLevel.INFO` a private constant, along with explanation of what they are. Perhaps the `true` in `includeAppLogs` also.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+        List<String> defaultVersions = new ArrayList<String>();
+        try {
+            currentVersionIndex = getCurrentVersionIndex(versionList, currentVersion);
+            defaultVersions = getNextFewVersions(versionList, currentVersionIndex);
```
I think it's safe for `currentVersionIndex` to be declared here instead.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+            defaultVersions = getNextFewVersions(versionList, currentVersionIndex);
+        } catch (InvalidParametersException e) {
+            defaultVersions.add(currentVersion);
+            e.printStackTrace();
```
`e.printStackTrace()` is hardly helpful. We can either don't have it, or do a `log.*` instead, although I'm not sure which log level is appropriate here.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+            if (versionList[i].equals(currentVersion)) return i;
+        }
+        throw new InvalidParametersException("The current version is not found!");
+    }
```
Doesn't `List<String> versionList` have a built-in function to achieve this?
To convert from `Set<String>` you just need to do this:
```java
Set<String> set = ... // get the set
List<String> list = new ArrayList<String>(set);
```

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,128 @@
+            }
+        }
+        return result;
+    }
```
Again, if you used `List<String>`, this is achieveable by `list.subList(fromIndex, toIndex)`.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -173,116 +165,39 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
+        LogHelper logHelper = new LogHelper();
+        logHelper.setQuery(data.getVersions(), data.getFromDate(), data.getToDate(), null);
+        
+        //LogQuery query = buildQuery(data);
```
Don't forget to remove this.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -59,123 +52,28 @@ protected ActionResult execute() throws EntityDoesNotExistException {
-            lastOffset = record.getOffset();
+        List<AppLogLine> appLogLines = logHelper.fetchLogs();
+        for (AppLogLine appLog : appLogLines) {
+            if (currentLogsInPage >= LOGS_PER_PAGE) break;
```
```java
if (condition) {
    break;
}
```

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -59,123 +52,28 @@ protected ActionResult execute() throws EntityDoesNotExistException {
+            if (currentLogsInPage >= LOGS_PER_PAGE) break;
+            String logMsg = appLog.getLogMessage();
+            boolean isNotEmailLog = (!logMsg.contains("TEAMMATESEMAILLOG"));
+            if (isNotEmailLog) continue;
```
Ditto here.

bamboo3250 commented:
@wkurniawan07 I did a few major changes.

1. Introduced VersionHelper and LogHelper.
2. Added another function to query logs with time increment in LogHelper. 
3. Changed logic in EmailLogPageAction. Previously it retrieved logs based on `offset` of the log. Now I changed it to query by time to make it similar with ActivityLogPage, which is more updated.
4. Added 2 more lines in the statusMessage of AdminActivityLogPage to keep track of all available versions and versions used to query.

![screenshot 2016-02-20 10 02 26](https://cloud.githubusercontent.com/assets/8078323/13197142/b136eaa8-d81f-11e5-8fb7-4f664e05ba81.png)

damithc commented:
When  introducing an `AbcHelper` class, see if you can have a `Abc` class instead and those helper methods can go as static methods in the `Abc' class. The latter is more OOP.

bamboo3250 commented:
I renamed `VersionHelper` to `Version` and added instance methods for `Version` class. 
Ready for review.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,195 @@
+
+public class LogHelper {
+    /**
+     * 6 past versions to query, including the current version and its 5 preceding versions.
```
6 ~~past~~ versions

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,195 @@
+    public void setQuery(List<String> versionsToQuery, Long startTime, Long endTime) {
+        query.includeAppLogs(INCLUDE_APP_LOG);
+        query.batchSize(BATCH_SIZE);
+        query.minLogLevel(MIN_LOG_LEVEL);
```
Will it be sensible to remove all these settings to the constructor?

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,195 @@
+     */
+    public void setEndTime(Long endTime) {
+        this.endTime = endTime;
+    }
```
Consider making this private?
Otherwise you might as well make the field `endTime` public.

wkurniawan07 commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,195 @@
+        } catch (InvalidParametersException e) {
+            defaultVersions.add(currentVersion.toStringForQuery());
+            Utils.getLogger().severe("The current version is not found: " + e.getMessage());
+            e.printStackTrace();
```
Two issues:
- Don't include `e.printStackTrace()`.
- The message "The current version is not found" is almost identical to what is set in the Exception. `.severe(e.getMessage())` should be sufficient.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,167 @@
+    private static ModulesService modulesService = ModulesServiceFactory.getModulesService();
+    private Integer major;  //  the first number
+    private Integer minor;  //  the second number
+    private String patch;   //  the rest of version
```
Must this be `String`?

bamboo3250 replied:
Yes, to cover the case when more than 3 numbers may be used to present version. For example 11.05.01.01, etc.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,167 @@
+            list = versionInString.split("-", 3);   // split into at most 3 parts
+        } else {
+            list = versionInString.split(".", 3);
+        }
```
```java
} else if (versionInString.contains(".")) {
    ...
} else {
    // throw some Exception/log.severe/Assumption.fail/...
}
```

bamboo3250 replied:
What if the version only has 1 number, i.e. only major version number?

wkurniawan07 replied:
Ah ok, I always thought our convention is there's at least two numbers.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,167 @@
+     */
+    public boolean equals(Object anotherVersion) {
+        return toString().equals(anotherVersion.toString());
+    }
```
Just a practice in Java, if you override `equals` you would've to override `hashCode` as well so consider doing that.

bamboo3250 replied:
`hashCode` function added below.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,167 @@
+            if (patch2 == null) {
+                return -1;
+            }
+            return -patch1.compareTo(patch2);
```
Looks very repetitive. Can the component-wise comparison be abstracted out into a common function?

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -11,33 +10,29 @@
     private static final int RELEVANT_LOGS_PER_PAGE = 50;
-    private static final int SEARCH_TIME_INCREMENT = 2*60*60*1000;  // two hours in millisecond
-    private static final int MAX_SEARCH_TIMES = 12;                 // maximum 1 day
+    private static final int MAX_SEARCH_TIMES = 12; // search maximum 12 times with time increment
```
Rather than hard-coding 12, consider `(24 * 60 * 60 * 1000) / LogHelper.SEARCH_TIME_INCREMENT`. It would be resilient should you need to change the latter value as well.

wkurniawan07 commented:
Added some suggestions.

bamboo3250 commented:
I accepted most of suggestions. Ready for review again.

wkurniawan07 commented:
Looks good now :+1:
Anyway, as a follow up to this, I suggest you take a look at `CompileLogs`, `CompileLogsServlet`, and `EntityModifiedLogsServlet`. `LogHelper` might be used in those classes too.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,192 @@
+import com.google.appengine.api.log.RequestLogs;
+import com.google.appengine.api.log.LogService.LogLevel;
+
+public class LogHelper {
```
Add header comment.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,192 @@
+    /**
+     * 6 versions to query, including the current version and its 5 preceding versions.
+     */
+    private static final int MAX_VERSIONS_TO_QUERY = 6;
```
This comment should be a description of the variable rather than the value. But you can value as an example to explain the variable.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,192 @@
+    /**
+     * Always includes application logs
+     */
+    private static final boolean INCLUDE_APP_LOG = true;
```
Same here.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,192 @@
+     */
+    private static final int BATCH_SIZE = 1000;
+    private static final LogLevel MIN_LOG_LEVEL = LogLevel.INFO;
+    public static final int SEARCH_TIME_INCREMENT = 2*60*60*1000;  // two hours in millisecond
```
This one needs to be public?

damithc replied:
`milliseconds`

bamboo3250 replied:
This constant is used in AdminActivityLogPage and AdminEmailLogPage to calculate `MAX_SEARCH_TIMES` so I leave it public.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,192 @@
+        query.includeAppLogs(INCLUDE_APP_LOG);
+        query.batchSize(BATCH_SIZE);
+        query.minLogLevel(MIN_LOG_LEVEL);
+    }
```
Um, what does this class actually represents? A Log query result? A log reader? If it is a pure helper class, one doesn't expect to see a constructor.

bamboo3250 replied:
It is more like a log reader so I think it should have a constructor with those default parameters for the query.

damithc replied:
I think you have two classes here: a `AdminLogReader` and `AdminLogQuery`. What do you think?

damithc replied:
What about this comment?

bamboo3250 replied:
I saw it just a few minutes ago and I'm still thinking about it. I look at this class as an entity, which I can provide with `versions`, ` startTime` and `endTime` then I can fetch the logs from GAE server without touching to the `LogQuery` inside. If you want to separate the LogQuery from it, the `AdminLogQuery` may be just the same as `LogQuery` with different default values. I think it's fine to include the query inside this class.

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,171 @@
+import com.google.appengine.api.modules.ModulesService;
+import com.google.appengine.api.modules.ModulesServiceFactory;
+
+public class Version {
```
Header comment?

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,171 @@
+     * The rest of version. It should be String to cover versions with more than 3 numbers.
+     * For example: 15.56.01.01
+     */
+    private String patch;
```
Perhaps you can write one detailed comment to explain all three variables?

bamboo3250 commented:
@damithc comments added. Ready for your review.

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,193 @@
+        if (minor != null) {
+            String minorInString = String.valueOf(minor);
+            if (minorInString.length() < 2) {
+                minorInString = "0" + minorInString;
```
This means `new Version("1-1").toString()` gives `1.01`? That's a bit odd isn't it? May be you should keep the three parts as strings (in fact the entire string) and convert to numbers on demand? Alternatively, can store the original string in addition to the three numbers?

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,193 @@
+            int patchComparisonResult = compareVersionString(v1.getPatchVersion(), v2.getPatchVersion());
+            return patchComparisonResult;
+        }
+    }
```
Shouldn't this be in the `Version` class?

damithc replied:
Never mind. my mistake.

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    private String getPatchVersion() {
+        return patch;
+    }
```
Why do we have private getters?

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,193 @@
+        }
+        Collections.sort(versionList, new VersionComparator());
+        return versionList;
+    }
```
This method looks like it belongs in a `LogReader` class or a `GaeApi` class. It's fine to do the processing of the version info in this class but reading it from GAE doesn't belong in this class.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,196 @@
+        if (endTime != null) {
+            query.endTimeMillis(endTime);
+        }
+    }
```
1. Why this method checks for null but `setEndTime` method below doesn't?
2. Shouldn't this method call the `setEndTime` method?

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,196 @@
+     * Retrieves all logs within 2 hours before the endTime.
+     * We can use it again to get logs from the next 2 hours.
+     * @return logs within 2 hours before endTime
+     */
```
There is no mention of `2 hours` in the code but the comment mentions `2 hours`

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -151,138 +174,71 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
+     * Filters logs that should be shown on Admin Activity Log Page.
+     */
+    private List<ActivityLogEntry> filterLogsForActivityLogPage(List<AppLogLine> appLogLines,
+                                                             AdminActivityLogPageData data) {
```
Indent three more spaces to align with `List` above?

damithc commented on src/main/webapp/jsp/adminEmailLog.jsp:
```diff
@@ -14,7 +14,7 @@
     <%-- this form is used to store parameters for ajaxloader only --%>
     <form id="ajaxLoaderDataForm">
-        <input type="hidden" name="offset" value="${data.offset}">
+        <input type="hidden" name="offset" value="">
```
Is this change intentional?

bamboo3250 replied:
Yes, now `offset` is not a part of `AdminEmailLogPageData` anymore (its value will be assigned via javascript) so I removed it.

damithc commented:
Added some more comments. I'll review again after you've revised accordingly. Coaxing out the right classes might need a few iterations.

bamboo3250 commented:
Now we have `GaeApi` class to retrieve the current version and all available versions of the app. 
I also added `Comparable` interface for `Version` class based on the previous Comparator and refactored the code according to your suggestions.
Ready for review.

damithc commented:
There is another major comment about splitting the `LogHelper` class.

wkurniawan07 commented on src/main/java/teammates/common/util/GaeApi.java:
```diff
@@ -0,0 +1,36 @@
+ * Calls API functions of Google AppEngine
+ */
+public class GaeApi {
+    private static ModulesService modulesService = ModulesServiceFactory.getModulesService();
```
If this is intended to be non-modifiable, make it `final`.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,152 @@
+    private String originalRepresentation;
+    private Integer major;
+    private Integer minor;
+    private String patch;
```
I just thought something: what if all the components are left as String? Saves the trouble of having to parse.

bamboo3250 replied:
then the version 4.44 will come before 10.44 but it should be 10.44 that goes first. The major part at least should be Integer because the length of the number may not be the same.

bamboo3250 replied:
Okay, I have a solution to make them String.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,152 @@
+            return -1;
+        }
+        return -num1.compareTo(num2);
+    }
```
And also saves the trouble of having duplicate comparators here.

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,152 @@
+     */
+    public String getPatchVersion() {
+        return patch;
+    }
```
Are the getters used outside the class? If they're not, I don't really see the need for getters because you can very simply call `version.major/minor/patch`.

bamboo3250 replied:
I think we may need them later. No harm to have them anyway.

wkurniawan07 replied:
No harm, but no need also. Your call :p

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -11,33 +10,34 @@
+     * The maximum number of times to retrieve logs with time increment.
+     * The current value will search up to 24h.
+     */
+    private static final int MAX_SEARCH_TIMES = 24 * 60 * 60 * 1000 / LogHelper.SEARCH_TIME_INCREMENT;
```
One thing you can do is also separate the 24-hr value to another private var, something like:
```java
private static final int MAX_SEARCH_PERIOD = 24 * 60 * 60 * 1000; // 24 hrs in milliseconds
private static final int MAX_SEARCH_TIMES = MAX_SEARCH_PERIOD / LogHelper.SEARCH_TIME_INCREMENT;
```

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -151,138 +175,71 @@ private void generateStatusMessage(AdminActivityLogPageData data, List<ActivityL
 
+    /**
+     * Retrieves enough logs within 24 hour.
+     */
```
That way you don't have to hard-code 24 hours here. You can make it refer to the variable that stores the 24 hours (or whatever value it is, if going to be changed).

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -46,154 +53,63 @@ protected ActionResult execute() throws EntityDoesNotExistException {
-        
-        if (currentLogsInPage >= LOGS_PER_PAGE) {   
-            status += "<button class=\"btn-link\" id=\"button_older\" onclick=\"submitFormAjax('" + lastOffset + "');\">Older Logs </button>";              
+            emailLogs.add(emailLogEntry);
```
This would look logically clearer:
```java
if (data.shouldShowLog(emailLogEntry)) {
    emailLogs.add(emailLogEntry);
}
```

wkurniawan07 commented on src/main/java/teammates/common/util/GaeApi.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Calls API functions of Google AppEngine
+ */
+public class GaeApi {
```
Actually, `GaeApi` might be too broad a term is the functions only involve versioning. Consider renaming it to something else that captures the versioning functionality?

bamboo3250 replied:
I think there is no need to rename it. We might need to other API functions to it later. Who know :stuck_out_tongue:

damithc replied:
`GaeAdminApi`?
Description: `A wrapper class for GAE application admin API. e.g. version management`

wkurniawan07 commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,168 @@
+        
+        if (versionInString.contains("rc")) {
+            int rcIndex = versionInString.indexOf("rc");
+            versionInString = versionInString.substring(0, rcIndex);
```
`versionInString.replace("rc", "")` will do the job

bamboo3250 replied:
(y)

bamboo3250 commented:
@damithc I removed the constructor for `LogHelper` and converted all of its instance methods to class method. Is this good enough? 
@wkurniawan07 Now `major` and `minor` are also String. I changed the logic to compare two versions a bit. Please check my solution.

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,144 @@
+/**
+ * Represents a version by 3 parts: major version, minor version and patch version.
+ */
+public class Version implements Comparable<Version> {
```
Write unit tests for this class.

damithc commented on src/main/java/teammates/common/util/LogHelper.java:
```diff
@@ -0,0 +1,193 @@
+    
+    private static LogQuery query;
+    private static Long endTime;
+    private static List<String> versionList;
```
Non-final variables in a fully static class indicates a design smell. In other words, this is a singleton class, implemented as a fully static class (not recommended). 
I still think you need a `LogReader` class that can take a `AdminLogQuery` object, execute it, and return the result. Alternatively, you can move the `LogReader` stuff (i.e. methods that depend on GAE api) into `GaeAdminApi` class. What's left would be the `AdminLogQuery` class.

damithc commented on src/main/java/teammates/common/util/GaeAdminApi.java:
```diff
@@ -0,0 +1,36 @@
+        return new Version(modulesService.getCurrentVersion());
+    }
+    
+}
```
If you make the methods non-static, the GaeAdminApi object can be replaced using a stub (i.e. dependency injection) which is better for testability. Fully static classes gets in the way of dependency injection.

bamboo3250 replied:
Should I create a `inst()` function in this case?

damithc replied:
No, we don't want singleton either. Singleton too is not a very good pattern.

damithc replied:
Just make it a normal class with an empty constructor.

bamboo3250 commented:
@damithc I separated the LogQuery from the LogHelper and created a wrapper class for it and also converted functions in `GaeAdminApi` to non-static functions.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,155 @@
+import com.google.appengine.api.log.LogQuery;
+import com.google.appengine.api.log.LogService.LogLevel;
+
+public class AdminLogQuery {
```
Header comment?

damithc commented on src/main/java/teammates/common/util/LogReader.java:
```diff
@@ -0,0 +1,56 @@
+        query.setEndTime(startTime - 1);
+        return logs;
+    }
+}
```
I think you can move these into `GaeAdminApi`. Alternatively, have two classes: `GaeVersionApi` and `GaeLogApi`. Also, make the methods non-static.

damithc commented on src/main/java/teammates/common/util/LogReader.java:
```diff
@@ -0,0 +1,56 @@
+            query.setEndTime(TimeHelper.now(0.0).getTimeInMillis());
+        }
+        Long startTime = query.getEndTime() - SEARCH_TIME_INCREMENT;
+        query.setTimePeriodForQuery(startTime, query.getEndTime());
```
The five lines above looks like it belongs in the `AdminLogQuery` class.

You can have only one `fetchLogs` method (instead of the current two), but manipulate the query object before calling the method. e.g.
```java
fetchLogs(query);
...
query.setQueryWindowBack(1); //set back by one hour
reader.fetchLogs(query);
```
Is that feasible?

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,158 @@
+        setTimePeriodForQuery(startTime, endTime);
+        versionList = getVersionIdsForQuery(versionsToQuery);
+        query.majorVersionIds(versionList);
+    }
```
Merge with constructor?

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,158 @@
+/**
+ * A wrapper class for LogQuery to retrieve logs from GAE server.
+ */
+public class AdminLogQuery {
```
Write unit tests for non-trivial methods.

bamboo3250 commented:
I added unit test for `AdminLogQuery` and remove `fetchLogsInNextHours` in `GaeLogApi`.Ready for review.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+        query.includeAppLogs(INCLUDE_APP_LOG);
+        query.batchSize(BATCH_SIZE);
+        query.minLogLevel(MIN_LOG_LEVEL);
+    }
```
No need for this?

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+    }
+    
+    /**
+     * Sets time period to query logs then moves endTime back by the same amount.
```
This comment is not clear.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+    
+    /**
+     * Sets values for query.
+     * If versionsToQuery is null or empty, default versions will be used instead.
```
Is there a specific behavior if any of the times are null?

bamboo3250 replied:
If startTime is null, all logs until endTime will be retrieved.
If endTime is null, all logs from startTime will be retrieved.
If both are null, all logs will be retrieved.
This is behaviour of LogQuery.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+    public void setQueryWindowBackward(long timeInMillis) {
+        if (getEndTime() == null) {
+            setEndTime(TimeHelper.now(0.0).getTimeInMillis());
+        }
```
Is there any need to keep the end time as null? If there isn't, it can be set in the constructor.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+     * @return a list of default versions for query.
+     */
+    private List<String> getDefaultVersionIdsForQuery() {
+        GaeVersionApi adminApi = new GaeVersionApi();
```
This dependency doesn't feel right. This class has no business knowing the `GaeVersionApi` class. The call site (i.e. probably the `Action` class) should get the relevant stuff from the api and pass to this class. 
In fact, this whole method feels like it should be in `GaeVersionApi` class.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+            versionListInString.add(version.toStringForQuery());
+        }
+        return versionListInString;
+    }
```
Move to `Version' class?

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,175 @@
+            return versionIndex;
+        }
+        throw new InvalidParametersException("The current version is not found!");
+    }
```
This too can go to `Version` class?. It's dealing with versions and not accessing internal data of this class.

damithc commented:
Added some more comments. Mostly to ensure classes are cohesive and code appears where they belong.

bamboo3250 commented:
Changes made. I'm ready for review.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,116 @@
+        if (isVersionSpecifiedInRequest) {
+            return versions;
+        }
+        GaeVersionApi versionApi = new GaeVersionApi();
```
Remove this dependency. May be default versions can be fed as a parameter in the constructor and stored inside from that point.

damithc commented on src/main/java/teammates/common/util/GaeVersionApi.java:
```diff
@@ -0,0 +1,80 @@
+     * Gets a few recent versions for log query.
+     * @return a list of default versions for query.
+     */
+    public List<String> getDefaultVersionIdsForLogQuery() {
```
A method should not be named based on the caller. A method can be called by any other class. Name it based on what the method does.

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,144 @@
+    /**
+     * Converts to String in format XX-XX-XXXX
+     */
+    public String toStringForQuery() {
```
`toStringWithDashes`? (to avoid naming method based on usage)

damithc commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -82,16 +86,21 @@ protected ActionResult execute() throws EntityDoesNotExistException{
         //This is used to parse the filterQuery. If the query is not parsed, the filter function would ignore the query
         data.generateQueryParameters(filterQuery);
         
+        boolean ifContinueFromPreviousSearch = (!data.isFromDateSpecifiedInQuery()) && (!searchTimeOffset.isEmpty());
```
`ifContinueFromPreviousSearch` --> `isContinueFromPreviousSearch`?

damithc commented on src/test/java/teammates/test/cases/common/VersionTest.java:
```diff
@@ -0,0 +1,43 @@
+        
+        version1 = new Version("15.09");
+        version2 = new Version("15.09.01");
+        assertTrue(version1.compareTo(version2) > 0);
```
I remember some special handling was done for `rc` but I can't see any test cases for that.

bamboo3250 commented:
I have addressed above issues.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,93 @@
+    
+    /**
+     * Sets the time period to query logs from endTime back to endTime - timeInMillis
+     * then moves endTime to right before the query's start time.
```
What's the intention of this method? It's not clear.

bamboo3250 replied:
Probably I will remove this function and move its code outside.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,93 @@
+    /**
+     * Sets values for query.
+     * 
+     * @param versionsToQuery 
```
This param needs more explanations.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,93 @@
+     * 
+     * @param versionsToQuery 
+     * @param startTime
+     * @param endTime
```
Can any param be null? If not, add assumptions?

damithc commented on src/main/java/teammates/common/util/Version.java:
```diff
@@ -0,0 +1,144 @@
+     * 
+     * It also support RC versions, which has "rc" appended at the end of the string.
+     * For example: 5rc, 4.55rc, 5.55.01rc
+     */
```
Good comment. Merge it with class header comment or constructor header comment.

damithc commented on src/main/java/teammates/common/util/GaeVersionApi.java:
```diff
@@ -0,0 +1,75 @@
+public class GaeVersionApi {
+    /**
+     * Default constructor.
+     */
```
Redundant comment

damithc commented:
In much better shape now. Only a few small things to tidy up.
Even the two action classes are no longer ugly, although not ideal yet. 
In a separate issue, you may want to take another pass at each of the action classes to refactor them and ensure good test coverage.

bamboo3250 commented:
I removed `setQueryWindowBackward()` function and manipulated the time period in action classes instead.

damithc commented:
I think the code belongs in the adminLogQuery class, but needs to be
packaged correctly.
>
>

-- 
Sent from a mobile device.

bamboo3250 commented:
Okay, I will try to bring it back.

bamboo3250 commented:
I bought it back and renamed it to `moveTimePeriodBackward()`.

damithc commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,114 @@
+    
+    /**
+     * Sets values for query.
+     * If startTime is null, it will be considered as 0.
```
What does 0 mean? forever?

bamboo3250 replied:
The time used in LogQuery is Unix time. https://en.wikipedia.org/wiki/Unix_time
0 means it will take logs after Thursday, 1 January 1970 (since forever).

bamboo3250 replied:
I will add comment for this

damithc commented on src/main/java/teammates/common/util/GaeVersionApi.java:
```diff
@@ -0,0 +1,72 @@
+ * A wrapper class for GAE application admin API. e.g. version management
+ */
+public class GaeVersionApi {
+    public GaeVersionApi() {
```
Add blank line above this?

damithc commented:
@wkurniawan07 anything more to add?

wkurniawan07 commented on src/main/java/teammates/common/util/AdminLogQuery.java:
```diff
@@ -0,0 +1,118 @@
+     */
+    public void setStartTime(Long startTimeParam) {
+        startTime = startTimeParam;
+    }
```
Do the setters need to be public?

bamboo3250 replied:
But it also doesn't make sense to have private a setter. I created it just in case we need it later.

wkurniawan07 replied:
Private setter is actually not uncommon (in public getter-private setter combo). Also, "just in case we need it later" [is not a good reason to implement it now](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

wkurniawan07 commented on src/main/java/teammates/common/util/GaeVersionApi.java:
```diff
@@ -0,0 +1,73 @@
+public class GaeVersionApi {
+    
+    public GaeVersionApi() {
+    }
```
I don't think empty constructors need to be defined

wkurniawan07 commented on src/main/java/teammates/common/util/GaeVersionApi.java:
```diff
@@ -0,0 +1,73 @@
+import com.google.appengine.api.modules.ModulesServiceFactory;
+
+/**
+ * A wrapper class for GAE application admin API. e.g. version management
```
No longer accurate, since the function is already narrowed down to version management.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -11,33 +10,43 @@
+     * Maximum number of versions to query.
+     * The current value will include the current version and its 5 preceding versions.
+     */
+    private static final int MAX_VERSIONS_TO_QUERY = 6;
```
Consider representing the value as `1 + 5`; redundancy aside, it's in line with the explanation.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminActivityLogPageAction.java:
```diff
@@ -101,7 +116,20 @@ protected ActionResult execute() throws EntityDoesNotExistException{
+        return versionApi.getMostRecentVersions(MAX_VERSIONS_TO_QUERY);
+    }
+    
+    private void generateStatusMessage(List<String> versionToQuery, AdminLogQuery query, AdminActivityLogPageData data, List<ActivityLogEntry> logs, String courseId) {
```
A line break here would be good.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -1,42 +1,59 @@
         
-        AdminEmailLogPageData data = new AdminEmailLogPageData(account, getRequestParamValue("offset"),
-                                        getRequestParamValue("filterQuery"), getRequestParamAsBoolean("all"));
+        AdminEmailLogPageData data = new AdminEmailLogPageData(account, getRequestParamValue("filterQuery"), getRequestParamAsBoolean("all"));
```
Line break here too.

wkurniawan07 commented on src/main/java/teammates/ui/controller/AdminEmailLogPageAction.java:
```diff
@@ -46,154 +63,82 @@ protected ActionResult execute() throws EntityDoesNotExistException {
-        return recentVersions;
+        String status="&nbsp;&nbsp;Total Logs gone through in last search: " + totalLogsSearched + "<br>";
+        //link for Next button, will fetch older logs
+        status += "<button class=\"btn-link\" id=\"button_older\" onclick=\"submitFormAjax('" + nextEndTimeToSearch + "');\">Search More</button>";           
```
And here.

wkurniawan07 commented:
Added some comments, mostly cosmetic changes.

bamboo3250 commented:
Cosmetic changes made. Ready for review.

bamboo3250 commented:
Okay, 2 setters are removed

wkurniawan07 commented:
Looks good now :+1:
Again, as a follow up to this, I suggest you take a look at `CompileLogs`, `CompileLogsServlet`, and `EntityModifiedLogsServlet` (all in `logic.automated`). The new classes defined here might be reusable in those places.
